### PR TITLE
MAINT: Use list comprehension when possible

### PR DESCRIPTION
**Pull request by rth**
This uses list comprehension when possible, instead of creating a list by appending to it in a for loop, which is faster and less verbose.

Tried to only make the changes in some simple cases when it doesn't hurt readability

### CODE REVIEW

File: scipy/io/idl.py [hunk 1]
```diff
@@ -395,17 +395,13 @@ def _read_record(f):
     elif record['rectype'] == "HEAP_HEADER":
 
         record['nvalues'] = _read_long(f)
-        record['indices'] = []
-        for i in range(record['nvalues']):
-            record['indices'].append(_read_long(f))
+        record['indices'] = [_read_long(f) for i in range(record['nvalues'])]
```
Comment (42B) [inline scipy/io/idl.py]: `i` isn't used, couldn't it be like this?

```python
[_read_long(f) for _ in range(record['nvalues'])]
```

File: scipy/io/idl.py [hunk 2]
```diff
@@ -395,17 +395,13 @@ def _read_record(f):
...
 
     elif record['rectype'] == "COMMONBLOCK":
 
         record['nvars'] = _read_long(f)
         record['name'] = _read_string(f)
-        record['varnames'] = []
-        for i in range(record['nvars']):
-            record['varnames'].append(_read_string(f))
+        record['varnames'] = [_read_string(f) for i in range(record['nvars'])]
```
Comment (42B) [inline scipy/io/idl.py]:
```python
[_read_string(f) for _ in range(record['nvars'])]  # ?
```

File: scipy/io/idl.py [hunk 3]
```diff
@@ -466,9 +462,7 @@ def _read_arraydesc(f):
 
         arraydesc['nmax'] = _read_long(f)
 
-        arraydesc['dims'] = []
-        for d in range(arraydesc['nmax']):
-            arraydesc['dims'].append(_read_long(f))
+        arraydesc['dims'] = [_read_long(f) for d in range(arraydesc['nmax'])]
```
Comment (42B) [inline scipy/io/idl.py]: `_`

File: scipy/io/idl.py [hunk 4]
```diff
@@ -518,32 +512,27 @@ def _read_structdesc(f):
 
     if not structdesc['predef']:
 
-        structdesc['tagtable'] = []
-        for t in range(structdesc['ntags']):
-            structdesc['tagtable'].append(_read_tagdesc(f))
+        structdesc['tagtable'] = [_read_tagdesc(f)
+                                  for t in range(structdesc['ntags'])]
```
Comment (42B) [inline scipy/io/idl.py]: `_`

File: scipy/io/idl.py [hunk 5]
```diff
@@ -518,32 +512,27 @@ def _read_structdesc(f):
...
             structdesc['classname'] = _read_string(f)
             structdesc['nsupclasses'] = _read_long(f)
-            structdesc['supclassnames'] = []
-            for s in range(structdesc['nsupclasses']):
-                structdesc['supclassnames'].append(_read_string(f))
-            structdesc['supclasstable'] = []
-            for s in range(structdesc['nsupclasses']):
-                structdesc['supclasstable'].append(_read_structdesc(f))
+            structdesc['supclassnames'] = [
+                _read_string(f) for s in range(structdesc['nsupclasses'])]
```
Comment (42B) [inline scipy/io/idl.py]: `_`

File: scipy/io/idl.py [hunk 6]
```diff
@@ -518,32 +512,27 @@ def _read_structdesc(f):
...
+            structdesc['supclasstable'] = [
+                _read_structdesc(f) for s in range(structdesc['nsupclasses'])]
```
Comment (42B) [inline scipy/io/idl.py]: `_`

File: scipy/optimize/_trustregion_constr/canonical_constraint.py [hunk 7]
```diff
@@ -99,12 +99,8 @@ def concatenate(cls, canonical_constraints, sparse_jacobian):
...
         """
         def fun(x):
-            eq_all = []
-            ineq_all = []
-            for c in canonical_constraints:
-                eq, ineq = c.fun(x)
-                eq_all.append(eq)
-                ineq_all.append(ineq)
+            eq_all, ineq_all = zip(
+                    *[c.fun(x) for c in canonical_constraints])
```
Comment (42B) [inline scipy/optimize/_trustregion_constr/canonical_constraint.py]: Couldn't this be on one line? If not, isn't the second line indented too far?

File: scipy/optimize/_trustregion_constr/canonical_constraint.py [hunk 8]
```diff
@@ -114,12 +110,8 @@ def fun(x):
...
 
         def jac(x):
-            eq_all = []
-            ineq_all = []
-            for c in canonical_constraints:
-                eq, ineq = c.jac(x)
-                eq_all.append(eq)
-                ineq_all.append(ineq)
+            eq_all, ineq_all = zip(
+                    *[c.jac(x) for c in canonical_constraints])
```
Comment (42B) [inline scipy/optimize/_trustregion_constr/canonical_constraint.py]: One line or less indentation?

**Comment by rth** [comment]
@42B Thanks. I addressed your comments.

**Comment by 42B** [comment]
@rth 

Looks good!

**Comment by rgommers** [comment]
> This one you mean #9531 ? :)

ah yes, indeed

**Comment by rth** [comment]
One change broke CI, will investigate and comment when CI is green.

File: scipy/spatial/qhull.pyx [hunk 9]
```diff
@@ -1009,10 +1007,8 @@ cdef void _visit_voronoi(qhT *_qh, void *ptr, vertexT *vertex, vertexT *vertexA,
     p[2*qh._nridges + 1] = point_2
 
     # Record which voronoi vertices constitute the ridge
-    cur_vertices = []
-    for i in xrange(qh_setsize(_qh, centers)):
-        ix = (<facetT*>centers.e[i].p).visitid - 1
-        cur_vertices.append(ix)
+    cur_vertices = [(<facetT*>centers.e[i].p).visitid - 1
+                    for i in xrange(qh_setsize(_qh, centers))]
```
Comment (rgommers) [inline scipy/spatial/qhull.pyx]: There's test failures on TravisCI for `voronoi` functionality that looks real, due to this change or the one above. Can you debug or revert?

Comment (rth) [reply scipy/spatial/qhull.pyx]: Yes, this like was breaking tests. I'm not sure what was wrong in the above change, so I reverted it.

**Comment by rth** [comment]
Tests should be mostly green,
 - pypy3 fails once, but it also does fail on master in that test,
   ```
   Worker 'gw2' crashed while running 'build/testenv/site-packages/scipy/sparse/linalg/isolve/tests/test_gcrotmk.py::TestGCROTMK::test_arnoldi'
   ```
 - Python 2.7 Travis CI fails with,
   ```
   scipy/special/_generate_pyx.py:841: list comprehension redefines '_' from line 833
   The command "PYFLAKES_NODOCTEST=1 pyflakes scipy benchmarks/benchmarks | grep -E -v 'unable to detect undefined names|assigned to but never used|imported but unused|redefinition of unused|may be undefined, or defined from star imports' > test.out; cat test.out; test \! -s test.out" exited with 1.
   ```
   which is a false postive IMO. This should have been fixed a few month ago in https://github.com/PyCQA/pyflakes/issues/202 but that CI has pyflake==1.1.0 pinned for some reason.

**Comment by rgommers** [comment]
> This should have been fixed a few month ago in PyCQA/pyflakes#202 but that CI has pyflake==1.1.0 pinned for some reason.

That was done in gh-5911 because it was flaky. We need to have it pinned, but if the latest released version of pyflakes works now, then updating to that version is fine. Can you try?

**Comment by rth** [comment]
Tried pinning `pyflake==2.0.0` which also works, but produces the same error. After further investigation, the corresponding fix didn't make it into 2.0.0. Reverted that commit as I am not sure that pyflake error messages are backward compatible and that `pyflakes | grep ....` works as expected -- probably better to do that in a separate PR.  

I ended up just adding a `# noqa` as a workaround for the false positive lint error.

**Comment by rth** [comment]
Fixed the pyflakes issue, but now one test is failing on Azure with a MemoryError,
```
2018-12-14T16:28:29.6315523Z     def test_arnoldi(self):
2018-12-14T16:28:29.6315961Z         np.random.rand(1234)
2018-12-14T16:28:29.6316147Z     
2018-12-14T16:28:29.6316676Z >       A = eye(10000) + rand(10000,10000,density=1e-4)
2018-12-14T16:28:29.6317030Z 
2018-12-14T16:28:29.6317654Z self       = <scipy.sparse.linalg.isolve.tests.test_gcrotmk.TestGCROTMK object at 0x0D441C50>
2018-12-14T16:28:29.6318098Z 
2018-12-14T16:28:29.6318360Z C:\hostedtoolcache\windows\Python\3.6.4\x86\lib\site-packages\scipy\sparse\linalg\isolve\tests\test_gcrotmk.py:64: 
2018-12-14T16:28:29.6319278Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2018-12-14T16:28:29.6319984Z C:\hostedtoolcache\windows\Python\3.6.4\x86\lib\site-packages\scipy\sparse\construct.py:842: in rand
2018-12-14T16:28:29.6320505Z     return random(m, n, density, format, dtype, random_state)
2018-12-14T16:28:29.6321143Z C:\hostedtoolcache\windows\Python\3.6.4\x86\lib\site-packages\scipy\sparse\construct.py:788: in random
2018-12-14T16:28:29.6321460Z     ind = random_state.choice(mn, size=k, replace=False)
2018-12-14T16:28:29.6321977Z mtrand.pyx:1190: in mtrand.RandomState.choice
2018-12-14T16:28:29.6322326Z     ???
2018-12-14T16:28:29.6322839Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2018-12-14T16:28:29.6322959Z 
2018-12-14T16:28:29.6323085Z >   ???
2018-12-14T16:28:29.6323785Z E   MemoryError
2018-12-14T16:28:29.6324052Z 
```

I don't think it's related to this PR?

**Comment by tylerjereddy** [comment]
> I don't think it's related to this PR?

Agreed, I've observed this myself lately -- I've raised a separate issue in #9595.

**Comment by rgommers** [comment]
Thanks @rth, merged.

**Pull request merged**